api/pkg/provider/cloud/vsphere/provider.go

```
@@ -118,92 +135,35 @@ func (v *Provider) InitializeCloudProvider(cluster *kubermaticv1.Cluster, update
 	// networks without entering credentials
-	if v.dc.InfraManagementUser != nil {
-		cloud.VSphere.InfraManagementUser.Username = v.dc.InfraManagementUser.Username
-		cloud.VSphere.InfraManagementUser.Password = v.dc.InfraManagementUser.Password
-	}
-
-	client, err := v.getClient(cloud)
+	session, err := v.newSession(ctx, cloud)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't initialize vsphere client: %v", err)
+		return nil, fmt.Errorf("failed to create vCenter session: %v", err)
 	}
-	defer logout(client)
-
-	finder := find.NewFinder(client.Client, true)
-
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	defer session.Logout()
 
-	vsphereDC, err := finder.Datacenter(ctx, v.dc.Datacenter)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get vsphere datacenter: %v", err)
-	}
-	finder.SetDatacenter(vsphereDC)
-
-	// finder is relative to datacenter, so * is fine for us.
-	netRefs, err := finder.NetworkList(ctx, "*")
-	if err != nil {
-		return nil, fmt.Errorf("couldn't retrieve network list: %v", err)
-	}
-
-	var networks []Network
-	for _, netRef := range netRefs {
-		backing, err := netRef.EthernetCardBackingInfo(ctx)
-		if err != nil {
-			// So, there are some netRefs (for example VmwareDistributedVirtualSwitch) which don't implement a BackingInfo.
-			// And since the error isn't typed, we can't check for it.
-			// And since the vsphere documentation is glorious, we dont know all netRef-types affected by this.
-			// so instead of checking type for ignoring that special error, we have to ignore all of it.
-			// return nil, fmt.Errorf("couldn't get network backing: %v (%s)", err, netRef.Reference().Type)
-			continue
-		}
-		netBacking, ok := backing.(*types.VirtualEthernetCardNetworkBackingInfo)
-		if !ok {
-			// ignore virtual networks
-			continue
-		}
-
-		network := Network{Name: netBacking.DeviceName}
-		networks = append(networks, network)
-	}
-
-	return networks, nil
+	return getPossibleVMNetworks(ctx, session)
```

This method here tries to get all the possible networks and it ignores all the errors which are reported in the process. Now, if all the tries failed the handler will handle the empty slice however it would be nice to have a general log message or potentially an error here, that none of the tries has succeeded.  
```
if info, err := getPossibleVMNetworks(ctx, session); err != nil || len(info) < 1 {
    return nil, errors.New("no possible networks were found")
}
```
Or simply add this check in the `getPossibleVMNetworks` method.

---

We directly return the error here - So in case something fails the caller will now ?

---

Not really you ignore the errors here: 
https://github.com/kubermatic/kubermatic/pull/4111/files#diff-7f41c6031dfdb241d336cde37220601dR32-R42

---

In case we decide to ignore an error there, we log it - which should be enough.
What would an additional logging help here?

---

If there was no network available([]NetworkInfo length is Zero) do you consider this an error or not? In your comment you say:

`// Just log the error. We do not want a single weird network to break the entire listing.`


No single network is available this means the entire network listing is broken. So instead of going to the logs and start counting how many broken networks we have and associating them with list NetworksList from the session Finder.

---

> If there was no network available([]NetworkInfo length is Zero) do you consider this an error or not?

An empty list is fine.

> No single network is available this means the entire network listing is broken. 

No it does not. An empty slice means, there are no VM networks available.
I'll update that comment to make that more clear